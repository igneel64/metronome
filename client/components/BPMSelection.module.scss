.bpmSelection {
    display: flex;
    flex-flow: wrap;
    justify-content: space-between;

    margin-bottom: 1.375rem;

    button {
        margin-bottom: 0.625rem;
    }
}