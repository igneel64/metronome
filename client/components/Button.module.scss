@use "../../styles/theme";

$button-background: #2d1c3f;
$button-color: #8eacff;
$active-button-background: #8eacff;
$active-button-color: #161020;

.button {
    background: $button-background;
    border-radius: 5px;
    box-shadow: 0px 1px 2px 0px rgba(0, 0, 0, 0.5);
    color: $button-color;
    font-size: theme.$small;
    padding: 0.625rem 0rem;
    width: 85px;
}

.active {
    background: $active-button-background;
    color: $active-button-color;
}