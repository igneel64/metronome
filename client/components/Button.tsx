import clsx from "clsx";

import styles from "./Button.module.scss";

type ButtonProps = {
  children: React.ReactNode;
  handleClick: () => void;
  isActive?: boolean;
};

export function Button({ children, handleClick, isActive }: ButtonProps) {
  return (
    <button
      onClick={handleClick}
      className={clsx(styles.button, isActive && styles.active)}
    >
      {children}
    </button>
  );
}
