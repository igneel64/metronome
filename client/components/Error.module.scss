@use '../../styles/theme';

.error {
    color: theme.$color-error;
    font-size: theme.$xxl;
    text-align: center;
    margin-bottom: 3rem;
}