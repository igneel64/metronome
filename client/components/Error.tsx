import styles from "./Error.module.scss";

export function Error() {
  return (
    <span className={styles.error} data-test-id="error">
      An error has occurred, please try again in a bit!
    </span>
  );
}
