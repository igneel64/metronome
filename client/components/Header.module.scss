@use '../../styles/theme';

.header {
    letter-spacing: 6px;
    font-weight: normal;
    font-size: theme.$large;
    margin-bottom: 3rem;
}