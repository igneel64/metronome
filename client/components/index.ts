export * from "./BPMSelection";
export * from "./Button";
export * from "./Header";
export * from "./Pulse";
export * from "./SimilarSongs";
export * from "./SpotifyRefresh";
