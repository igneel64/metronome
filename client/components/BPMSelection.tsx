import { useEffect, useMemo, useState } from "react";

import { useMetronome, useSongs } from "../hooks";
import { ascendingOrderNumeric } from "../utils/ascendingOrderNumeric";
import styles from "./BPMSelection.module.scss";
import { Button } from "./Button";
import { Error } from "./Error";
import { Loader } from "./Loader";

const INACTIVE_INDEX = -1;

export function BPMSelection() {
  const [active, setActive] = useState(INACTIVE_INDEX);
  const { setBpm } = useMetronome();
  const { songs, loading, hasError } = useSongs();

  const bpmSet = useMemo(() => uniqueBpmSet(songs), [songs]);

  useEffect(() => setActive(INACTIVE_INDEX), [songs]);

  const handleBPMSelection = (bpm: number, idx: number) => {
    setBpm(bpm);
    setActive(idx);
  };

  if (loading) {
    return <Loader />;
  }

  if (hasError) {
    return <Error />;
  }

  return (
    <div className={styles.bpmSelection} data-test-id="select-bpm">
      {[...bpmSet].map((bpm, idx) => (
        <Button
          key={`${bpm}-${idx}`}
          handleClick={() => handleBPMSelection(bpm, idx)}
          isActive={idx === active}
        >
          {bpm} BPM
        </Button>
      ))}
    </div>
  );
}

function uniqueBpmSet(songs: (string | number)[][]) {
  return new Set(
    songs.map(([_, __, bpm]) => Number(bpm)).sort(ascendingOrderNumeric)
  );
}
