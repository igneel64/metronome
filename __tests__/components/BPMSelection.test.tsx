import { render } from "@testing-library/react";

import { BPMSelection } from "../../client/components/BPMSelection";

describe("BPMSelection", () => {
  it("renders properly", () => {
    const { container } = render(<BPMSelection />);
    expect(container).toMatchSnapshot();
  });
});
