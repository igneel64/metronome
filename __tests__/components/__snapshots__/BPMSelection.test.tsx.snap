// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BPMSelection renders properly 1`] = `
<div>
  <div
    class="bpmSelection"
  >
    <button
      class="button"
    >
      72
       BPM
    </button>
    <button
      class="button"
    >
      74
       BPM
    </button>
    <button
      class="button"
    >
      82
       BPM
    </button>
    <button
      class="button"
    >
      84
       BPM
    </button>
    <button
      class="button"
    >
      87
       BPM
    </button>
    <button
      class="button"
    >
      92
       BPM
    </button>
    <button
      class="button"
    >
      96
       BPM
    </button>
    <button
      class="button"
    >
      128
       BPM
    </button>
    <button
      class="button"
    >
      138
       BPM
    </button>
  </div>
</div>
`;
